File: apps/backend-go/pkg/security/password.go
package security

import (
	"errors"
	"regexp"
	"unicode"

	"golang.org/x/crypto/bcrypt"
)

// Password validation constants
const (
	MinPasswordLength = 8
	MaxPasswordLength = 128
	BcryptCost        = 12 // Recommended by OWASP
)

var (
	ErrPasswordTooShort   = errors.New("password must be at least 8 characters")
	ErrPasswordTooLong    = errors.New("password must not exceed 128 characters")
	ErrPasswordTooWeak    = errors.New("password must contain uppercase, lowercase, number, and special character")
	ErrInvalidPassword    = errors.New("invalid password")
)

// PasswordStrength represents the strength of a password
type PasswordStrength struct {
	Score    int
	Feedback []string
}

// HashPassword hashes a password using bcrypt
func HashPassword(password string) (string, error) {
	// Validate password strength
	if err := ValidatePassword(password); err != nil {
		return "", err
	}

	hashedBytes, err := bcrypt.GenerateFromPassword([]byte(password), BcryptCost)
	if err != nil {
		return "", err
	}

	return string(hashedBytes), nil
}

// VerifyPassword verifies a password against a hash
func VerifyPassword(password, hash string) bool {
	err := bcrypt.CompareHashAndPassword([]byte(hash), []byte(password))
	return err == nil
}

// ValidatePassword validates password strength
func ValidatePassword(password string) error {
	// Length check
	if len(password) < MinPasswordLength {
		return ErrPasswordTooShort
	}
	if len(password) > MaxPasswordLength {
		return ErrPasswordTooLong
	}

	// Character diversity check
	var (
		hasUpper   bool
		hasLower   bool
		hasNumber  bool
		hasSpecial bool
	)

	for _, char := range password {
		switch {
		case unicode.IsUpper(char):
			hasUpper = true
		case unicode.IsLower(char):
			hasLower = true
		case unicode.IsNumber(char):
			hasNumber = true
		case unicode.IsPunct(char) || unicode.IsSymbol(char):
			hasSpecial = true
		}
	}

	if !hasUpper || !hasLower || !hasNumber || !hasSpecial {
		return ErrPasswordTooWeak
	}

	return nil
}

// CalculatePasswordStrength calculates the strength score of a password
func CalculatePasswordStrength(password string) PasswordStrength {
	score := 0
	feedback := []string{}

	// Length checks
	if len(password) >= 8 {
		score += 20
	}
	if len(password) >= 12 {
		score += 10
	}
	if len(password) >= 16 {
		score += 10
	}

	// Character diversity
	var hasUpper, hasLower, hasNumber, hasSpecial bool
	for _, char := range password {
		switch {
		case unicode.IsUpper(char):
			hasUpper = true
		case unicode.IsLower(char):
			hasLower = true
		case unicode.IsNumber(char):
			hasNumber = true
		case unicode.IsPunct(char) || unicode.IsSymbol(char):
			hasSpecial = true
		}
	}

	if hasUpper {
		score += 15
	} else {
		feedback = append(feedback, "Add uppercase letters")
	}

	if hasLower {
		score += 15
	} else {
		feedback = append(feedback, "Add lowercase letters")
	}

	if hasNumber {
		score += 15
	} else {
		feedback = append(feedback, "Add numbers")
	}

	if hasSpecial {
		score += 15
	} else {
		feedback = append(feedback, "Add special characters")
	}

	// Check for common passwords
	commonPasswords := []string{"password", "12345678", "qwerty", "admin", "letmein"}
	for _, common := range commonPasswords {
		if regexp.MustCompile("(?i)" + common).MatchString(password) {
			score -= 30
			feedback = append(feedback, "Avoid common passwords")
			break
		}
	}

	// Ensure score is between 0-100
	if score < 0 {
		score = 0
	}
	if score > 100 {
		score = 100
	}

	return PasswordStrength{
		Score:    score,
		Feedback: feedback,
	}
}

File: apps/backend-node/src/utils/security/password.ts
import bcrypt from "bcryptjs";

/**
 * Password Security Utilities for Node.js Backend
 * Provides secure password hashing, validation, and strength checking
 */

const BCRYPT_ROUNDS = 12; // OWASP recommended

export interface PasswordStrength {
  score: number;
  feedback: string[];
}

/**
 * Hash a password using bcrypt
 * @param password - Plain text password
 * @returns Hashed password
 */
export async function hashPassword(password: string): Promise<string> {
  // Validate password first
  const validation = validatePassword(password);
  if (!validation.valid) {
    throw new Error(validation.errors.join(", "));
  }

  return bcrypt.hash(password, BCRYPT_ROUNDS);
}

/**
 * Verify a password against a hash
 * @param password - Plain text password
 * @param hash - Hashed password
 * @returns True if password matches
 */
export async function verifyPassword(
  password: string,
  hash: string
): Promise<boolean> {
  try {
    return await bcrypt.compare(password, hash);
  } catch (error) {
    return false;
  }
}

/**
 * Validate password requirements
 * @param password - Password to validate
 * @returns Validation result
 */
export function validatePassword(password: string): {
  valid: boolean;
  errors: string[];
} {
  const errors: string[] = [];

  if (password.length < 8) {
    errors.push("Password must be at least 8 characters");
  }

  if (password.length > 128) {
    errors.push("Password must not exceed 128 characters");
  }

  if (!/[a-z]/.test(password)) {
    errors.push("Password must contain at least one lowercase letter");
  }

  if (!/[A-Z]/.test(password)) {
    errors.push("Password must contain at least one uppercase letter");
  }

  if (!/[0-9]/.test(password)) {
    errors.push("Password must contain at least one number");
  }

  if (!/[^a-zA-Z0-9]/.test(password)) {
    errors.push("Password must contain at least one special character");
  }

  return {
    valid: errors.length === 0,
    errors,
  };
}

/**
 * Calculate password strength
 * @param password - Password to check
 * @returns Strength score and feedback
 */
export function calculatePasswordStrength(
  password: string
): PasswordStrength {
  let score = 0;
  const feedback: string[] = [];

  // Length checks
  if (password.length >= 8) score += 20;
  if (password.length >= 12) score += 10;
  if (password.length >= 16) score += 10;

  // Character diversity
  if (/[a-z]/.test(password)) {
    score += 15;
  } else {
    feedback.push("Add lowercase letters");
  }

  if (/[A-Z]/.test(password)) {
    score += 15;
  } else {
    feedback.push("Add uppercase letters");
  }

  if (/[0-9]/.test(password)) {
    score += 15;
  } else {
    feedback.push("Add numbers");
  }

  if (/[^a-zA-Z0-9]/.test(password)) {
    score += 15;
  } else {
    feedback.push("Add special characters");
  }

  // Common password check
  const commonPasswords = ["password", "12345678", "qwerty", "admin"];
  if (commonPasswords.some((p) => password.toLowerCase().includes(p))) {
    score = Math.max(0, score - 30);
    feedback.push("Avoid common passwords");
  }

  return {
    score: Math.min(100, score),
    feedback,
  };
}
